proposals/002-text-utf-default.md

```diff
@@ -0,0 +1,307 @@
+a character (as of Unicode 13.0). Now comes a tricky part: Unicode defines how to map
+characters to code points (basically, integers), but how would you serialise lists of
+integers to bytes? The simplest encoding is just allocate 32 bits (4 bytes) per code
+point, and write them one by one. This is UTF-32. Its main benefit is that since
+all code points take the same size, you can still index characters in a constant time.
+However, memory requirements are 4x comparing to ASCII, and in a world of ASCII and UCS-2
+there was little appetite to embrace one more incompatible encoding.
+
+Next option on the list is to encode some code points as 2 bytes and some others,
+less lucky ones, as 4 bytes. This is UTF-16. This encoding allowed to retain a decent
+backward compatibility with UCS-2, so, for instance, modern Java and JavaScript
+stick to UTF-16 as default internal representation. The biggest downside is that you
+no longer know beforehand, where *n*-th code point starts. One must *parse* all characters
+from the very beginning to learn which ones are 2-byte long and which are 4-byte long.
+
+But once we abandon requirement of constant indexing, even better option arises. Let's
+encode first 128 characters as 1 byte, some others as 2 bytes, and the rest as 3 or 4 bytes.
+This is UTF-8. The killer feature of this encoding is that it's fully backwards compatible
+with ASCII. This meant that all existing ASCII documents were automatically valid UTF-8
+documents as well, and that 50-years-old executables could often parse UTF-8 data without
+knowing a bit about it. This property appeared so important that in a modern environment
+the vast majority of data is stored and sent between processes in UTF-8 encoding.
+
+To sum up:
+
+* Both UTF-8 and UTF-16 support exactly the same range of characters.
+* For ASCII data UTF-8 takes twice less space.
+* UTF-8 is vastly more popular for serialization and storage.
+
+# Motivation
+
+`text` is a standard Haskell library for Unicode strings. Internally it stores
+Unicode code points in UTF-16, so any character takes either 2 or 4 bytes.
+In a modern enviroment this is a suboptimal choice: usually data
+is stored (e. g., on a disc or in DB) and tranferred between agents (e. g., via web)
+in UTF-8 encoding. So `text` needs to convert (UTF-8 to UTF-16) all inputs and usually
+ends up converting outputs as well (this time UTF-16 to UTF-8).
+
+Even within Haskell ecosystem UTF-16 is rarely used
+for interprocess communication or as a component of binary formats.
+The very `instance Binary Text` serializes `Text` in UTF-8 encoding.
+
+If we switch the internal representation of `Text` from UTF-16 to UTF-8,
+all such conversions would be made redundant and we'll be able just check that
+a `ByteString` is a valid UTF-8 (which is most often the case) and copy it into `Text`.
+If in future (see an upcoming "Unifying vector-like types" proposal) `ByteString`
```

Comment:
I don't like this reference. it sounds as "Unifying vector-like types" is already agreed upon. That is not true.

Reply:
I don't feel it says so, it clearly starts with "If". I'd like to put this proposal in a context of other HF developments. I'll improve wording in the next iteration.

Reply:
I'm worried about the exact content of that proposal. As far as I understood it degraded to something as moderate as "add `primitive` types to `base` and see then what next". I.e. nothing concrete about changing `bytestring` internal representation (it's just changed, not fun changing it again soon, is anyone using `bytestring-0.11` in production?).

Reply:
We digress, but as you might guess I'm not a fan of changing `bytestring` internal representation as well, `0.11` migration was exhausting. My personal attitude is stated at https://github.com/haskell/bytestring/issues/193#issuecomment-770262177